lib/enigma.rb

```diff
@@ -1,2 +1,121 @@
 class Enigma
+  attr_reader :characters
+  def initialize
+    @characters = ("a".."z").to_a << " "
+  end
+
+  def generate_key_string
+    rand(99999).to_s.rjust(5, "0")
+  end
+
+  def generate_keys(key_string)
+    {
+    a: key_string[0, 2].to_i,
+    b: key_string[1, 2].to_i,
+    c: key_string[2, 2].to_i,
+    d: key_string[3, 2].to_i
+  }
+  end
+
+  def generate_date
+    Time.now.strftime("%d%m%y")
+  end
+
+  def generate_offsets(date = generate_date)
```

Comment:
This method generates the offsets to be added to the initial shifts. This will be extracted to a `Shift` class.